xarray/core/indexes.py

```diff
@@ -387,103 +827,392 @@ def query(self, labels, method=None, tolerance=None):
+        "_dims",
+        "__coord_name_id",
+        "__id_index",
+        "__id_coord_names",
+    )
 
-    def __init__(self, indexes: Mapping[Any, Union[pd.Index, Index]]) -> None:
-        """Not for public consumption.
+    def __init__(
+        self,
+        indexes: dict[Any, T_PandasOrXarrayIndex],
+        variables: dict[Any, Variable],
+    ):
+        """Constructor not for public consumption.
 
         Parameters
         ----------
-        indexes : Dict[Any, pandas.Index]
+        indexes : dict
             Indexes held by this object.
+        variables : dict
+            Indexed coordinate variables in this object.
+
         """
         self._indexes = indexes
+        self._variables = variables
+
+        self._dims: Mapping[Hashable, int] | None = None
+        self.__coord_name_id: dict[Any, int] | None = None
+        self.__id_index: dict[int, T_PandasOrXarrayIndex] | None = None
+        self.__id_coord_names: dict[int, tuple[Hashable, ...]] | None = None
+
+    @property
+    def _coord_name_id(self) -> dict[Any, int]:
+        if self.__coord_name_id is None:
+            self.__coord_name_id = {k: id(idx) for k, idx in self._indexes.items()}
+        return self.__coord_name_id
+
+    @property
+    def _id_index(self) -> dict[int, T_PandasOrXarrayIndex]:
+        if self.__id_index is None:
+            self.__id_index = {id(idx): idx for idx in self.get_unique()}
+        return self.__id_index
+
+    @property
+    def _id_coord_names(self) -> dict[int, tuple[Hashable, ...]]:
+        if self.__id_coord_names is None:
+            id_coord_names: Mapping[int, list[Hashable]] = defaultdict(list)
+            for k, v in self._coord_name_id.items():
+                id_coord_names[v].append(k)
+            self.__id_coord_names = {k: tuple(v) for k, v in id_coord_names.items()}
+
+        return self.__id_coord_names
+
+    @property
+    def variables(self) -> Mapping[Hashable, Variable]:
+        return Frozen(self._variables)
+
+    @property
+    def dims(self) -> Mapping[Hashable, int]:
+        from .variable import calculate_dimensions
+
+        if self._dims is None:
+            self._dims = calculate_dimensions(self._variables)
+
+        return Frozen(self._dims)
+
+    def get_unique(self) -> list[T_PandasOrXarrayIndex]:
+        """Return a list of unique indexes, preserving order."""
+
+        unique_indexes: list[T_PandasOrXarrayIndex] = []
+        seen: set[T_PandasOrXarrayIndex] = set()
+
+        for index in self._indexes.values():
+            if index not in seen:
+                unique_indexes.append(index)
+                seen.add(index)
+
+        return unique_indexes
+
+    def is_multi(self, key: Hashable) -> bool:
+        """Return True if ``key`` maps to a multi-coordinate index,
+        False otherwise.
+        """
+        return len(self._id_coord_names[self._coord_name_id[key]]) > 1
+
+    def get_all_coords(
+        self, key: Hashable, errors: str = "raise"
```

Comment:
```suggestion
        self, key: Hashable, errors: Literal["raise", "ignore"] = "raise"
```
Can use Literal here and then I think we have done enough and can remove the runtime checks.